Mac support for extended proofreading.

megangardner opened the pull request:
#### 9840948a2ed4bca3b59b39f320610bff4e0133c4
<pre>
Mac support for extended proofreading.
<a href="https://bugs.webkit.org/show_bug.cgi?id=302917">https://bugs.webkit.org/show_bug.cgi?id=302917</a>
<a href="https://rdar.apple.com/165075699">rdar://165075699</a>

Reviewed by Aditya Keerthi and Abrar Rahman Protyasha.

Add support for extended proofreading.

Test: editing/input/cocoa/extended-proofreading.html
* LayoutTests/editing/input/cocoa/extended-proofreading-expected.txt: Added.
* LayoutTests/editing/input/cocoa/extended-proofreading.html: Added.
* LayoutTests/platform/ios/TestExpectations:
* Source/WebCore/PAL/pal/spi/mac/NSSpellCheckerSPI.h:
* Source/WebCore/editing/Editor.cpp:
(WebCore::Editor::markAllMisspellingsAndBadGrammarInRanges):
* Source/WebCore/editing/SpellChecker.cpp:
(WebCore::SpellCheckRequest::didSucceed):
(WebCore::SpellCheckRequest::setExistingResults):
(WebCore::SpellChecker::requestExtendedCheckingFor):
(WebCore::containsGrammarResult):
(WebCore::containsAdditionalGrammarResults):
(WebCore::SpellChecker::didCheck):
(WebCore::SpellChecker::didCheckSucceed):
(WebCore::SpellChecker::didCheckCancel):
* Source/WebCore/editing/SpellChecker.h:
* Source/WebCore/loader/EmptyClients.cpp:
(WebCore::EmptyEditorClient::EmptyTextCheckerClient::requestExtendedCheckingOfString):
* Source/WebCore/platform/text/TextCheckerClient.h:
* Source/WebKit/UIProcess/TextChecker.cpp:
(WebKit::TextChecker::requestExtendedCheckingOfString):
* Source/WebKit/UIProcess/TextChecker.h:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::requestExtendedCheckingOfString):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebPageProxy.messages.in:
* Source/WebKit/UIProcess/ios/TextCheckerIOS.mm:
(WebKit::TextChecker::requestExtendedCheckingOfString):
* Source/WebKit/UIProcess/mac/TextCheckerMac.mm:
(WebKit::convertExtendedCheckingResults):
(WebKit::TextChecker::requestExtendedCheckingOfString):
* Source/WebKit/WebProcess/WebCoreSupport/WebEditorClient.cpp:
(WebKit::WebEditorClient::requestExtendedCheckingOfString):
* Source/WebKit/WebProcess/WebCoreSupport/WebEditorClient.h:
* Source/WebKitLegacy/mac/WebCoreSupport/WebEditorClient.h:
* Source/WebKitLegacy/mac/WebCoreSupport/WebEditorClient.mm:
(WebEditorClient::requestExtendedCheckingOfString):
* Tools/TestRunnerShared/cocoa/LayoutTestSpellChecker.mm:
(-[LayoutTestSpellChecker requestGrammarCheckingOfString:range:language:options:completionHandler:]):

Canonical link: <a href="https://commits.webkit.org/303913@main">https://commits.webkit.org/303913@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/67c10df276c4487b029f26905bb95b48cf1167f7

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/133931 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/6442 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/45128 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/141511 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/85992 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/e076c5eb-1c53-4817-918b-483757161f70) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/135801 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/6973 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/6306 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/102447 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/69753 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/e705181b-40bf-495f-8d93-b4a16af64804) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/136878 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/162/builds/4902 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/120066 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/83246 "Passed tests") | | ⏳ 🛠 vision-apple 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/4779 "Passed tests") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/2397 "Passed tests") | [⏳ 🛠 wpe-cairo ](https://ews-build.webkit.org/#/builders/WPE-Cairo-Build-EWS "Waiting in queue, processing has not started yet") | | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/126008 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/113940 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/38186 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/144156 "Built successfully") | | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/132445 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/6112 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/38765 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/110813 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/6194 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/5188 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/111022 "Passed tests") | | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/4633 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/116324 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/59859 "Built successfully") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/20700 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/6164 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/34582 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/165408 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/6010 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/69628 "Built successfully") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/43197 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/6255 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/6118 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

aprotyas commented on Source/WebCore/editing/Editor.cpp:
```diff
@@ -3208,6 +3208,19 @@ void Editor::markAllMisspellingsAndBadGrammarInRanges(OptionSet<TextCheckingType
+    });
+
+    if (shouldMarkGrammar && extendedProofreadingEnabled) {
+        auto extendedRequest = SpellCheckRequest::create(resolvedOptions, TextCheckingProcessType::TextCheckingProcessIncremental, checkingRange, textReplacementRange, paragraphRange);
```
```suggestion
        RefPtr extendedRequest = SpellCheckRequest::create(resolvedOptions, TextCheckingProcessType::TextCheckingProcessIncremental, checkingRange, textReplacementRange, paragraphRange);
```

aprotyas commented on Source/WebCore/editing/SpellChecker.cpp:
```diff
@@ -210,12 +228,45 @@ void SpellChecker::enqueueRequest(Ref<SpellCheckRequest>&& request)
+        }
+        if (detailsMatch)
+            foundIt = true;
+    }
```
nit: A few suggestions...

It seems like `i` is only needed to access the i-th element, so we could do a range-for outer loop here.

Similarly for the inner loop, we can do a zipped range, and we change the loop content slightly to adjust for `detailsMatch`?

Also, I think we can remove some conditionals.

```suggestion
    for (TextCheckingResult existingResult : existingResults) {
        TextCheckingType existingResultType = *existingResult.type.toSingleValue();
        if (existingResultType != TextCheckingType::Grammar || result.range.location != existingResult.range.location || result.range.length != existingResult.range.length || result.details.size() != existingResult.details.size())
            continue;
        bool detailsMatch = true;
        for (auto [detail, existingResultDetail] : zippedRange(result.details, existingResult.details)) {
            if (!detailsMatch)
                break;
            detailsMatch = std::tie(detail.range.location, detail.range.length, detail.guesses) == std::tie(existingResultDetail.range.location, existingResultDetail.range.length, existingResultDetail.guesses);
        }
        if (detailsMatch)
            foundIt = true;
    }
```

aprotyas commented on Source/WebCore/editing/SpellChecker.cpp:
```diff
@@ -210,12 +228,45 @@ void SpellChecker::enqueueRequest(Ref<SpellCheckRequest>&& request)
+        if (resultType == TextCheckingType::Grammar && !containsGrammarResult(results[i], existingResults))
+            retval = true;
+    }
+    return retval;
```
nit: slight rewrite to avoid the retval on stack
```suggestion
    for (const auto& result : results) {
        TextCheckingType resultType = result.type.toSingleValue();
        if (resultType == TextCheckingType::Grammar && !containsGrammarResult(results[i], existingResults))
            return true;
    }
    return false;
```

aprotyas commented on Source/WebKit/UIProcess/mac/TextCheckerMac.mm:
```diff
@@ -551,6 +551,47 @@ static bool shouldSmartListsBeEnabled()
+
+void TextChecker::requestExtendedCheckingOfString(Ref<TextCheckerCompletion>&& textCheckerCompletion, int32_t insertionPoint)
+{
+    RetainPtr<NSString> textString = textCheckerCompletion->textCheckingRequestData().text().createNSString();
```
```suggestion
    RetainPtr textString = textCheckerCompletion->textCheckingRequestData().text().createNSString();
```

aprotyas reviewed (commented):
A few comments, but this looks mostly good to go!

aprotyas commented on Source/WebCore/editing/SpellChecker.cpp:
```diff
@@ -210,12 +228,44 @@ void SpellChecker::enqueueRequest(Ref<SpellCheckRequest>&& request)
-        m_requestQueue.clear();
+    bool foundIt = false;
+    for (TextCheckingResult existingResult : existingResults) {
+        TextCheckingType existingResultType = *existingResult.type.toSingleValue();
```
do we know for sure that `existingResult.type` has exactly one bit set? The optional unwrapping seems risky unless we know that, right?

aprotyas replied:
I would maybe replace this part with

```cpp
for (TextCheckingResult existingResult : existingResults) {
    if (!existingResult.type.containsOnly({ TextCheckingType::Grammar }) || ...)
```

to circumvent the force unwrap.

megangardner replied:
That does seem lots better. This is mostly a patch from Doug that I've tried to polish up, so I'm not sure if we can guarantee that there's only one type or not.

aprotyas replied:
Right, I would go with `containsOnly` just to minimize risk

aprotyas commented on Source/WebCore/editing/SpellChecker.cpp:
```diff
@@ -210,12 +228,44 @@ void SpellChecker::enqueueRequest(Ref<SpellCheckRequest>&& request)
+        TextCheckingType existingResultType = *existingResult.type.toSingleValue();
+        if (existingResultType != TextCheckingType::Grammar || result.range.location != existingResult.range.location || result.range.length != existingResult.range.length || result.details.size() != existingResult.details.size())
+            continue;
+        bool detailsMatch = true;
+        for (auto [detail, existingResultDetail] : zippedRange(result.details, existingResult.details)) {
+            if (!detailsMatch)
+                break;
+            detailsMatch = std::tie(detail.range.location, detail.range.length, detail.guesses) == std::tie(existingResultDetail.range.location, existingResultDetail.range.length, existingResultDetail.guesses);
+        }
```
totally optional, but if you would like to adopt it, then `ranges::all_of` is an idiom that expresses this pattern (even the early return!)

```suggestion
        bool detailsMatch = std::ranges::all_of(zippedRange(result.details, existingResult.details), [](const auto& detailPair) {
            auto [detail, existingResultDetail] = detailPair;
            return std::tie(detail.range.location, detail.range.length, detail.guesses) == std::tie(existingResultDetail.range.location, existingResultDetail.range.length, existingResultDetail.guesses);
        });
```

aprotyas replied:
no guarantees that it works with `WTF::ZipperedRange`, but here's an example with the standard library analogue https://godbolt.org/z/rad5WMGch

aprotyas commented on Source/WebCore/editing/SpellChecker.cpp:
```diff
@@ -210,12 +228,44 @@ void SpellChecker::enqueueRequest(Ref<SpellCheckRequest>&& request)
+{
+    for (const auto& result : results) {
+        TextCheckingType resultType = *result.type.toSingleValue();
+        if (resultType == TextCheckingType::Grammar && !containsGrammarResult(result, existingResults))
```
Same question about if we know for sure that `result.type` has exactly one bit set.

Maybe we can replace this in the same manner?

```suggestion
        if (result.type.containsOnly({ TextCheckingType::Grammar }) && !containsGrammarResult(result, existingResults))
```

aprotyas commented on Source/WebCore/PAL/pal/spi/mac/NSSpellCheckerSPI.h:
I suppose the changes in this file will be moot after your other patch lands.

megangardner replied:
Yeah, I just had this here cause I think not having it right was causing issues with the other checker.

aprotyas commented on Source/WebKit/UIProcess/mac/TextCheckerMac.mm:
```diff
@@ -551,6 +551,47 @@ static bool shouldSmartListsBeEnabled()
+{
+    RetainPtr textString = textCheckerCompletion->textCheckingRequestData().text().createNSString();
+    NSRange range = NSMakeRange(0, textCheckerCompletion->textCheckingRequestData().text().length());
+    auto textCompletion = WTFMove(textCheckerCompletion);
```
do we need to do this move? Can we not directly move `textCheckerCompletion` over in line 590? Something like this

```cpp
callOnMainRunLoop([completion = WTFMove(textCheckerCompletion), results] {
    completion->didFinishCheckingText(result);
}
```

aprotyas commented on Tools/TestRunnerShared/cocoa/LayoutTestSpellChecker.mm:
```diff
@@ -305,6 +306,16 @@ - (NSInteger)requestCheckingOfString:(NSString *)stringToCheck range:(NSRange)ra
 
+- (NSInteger)requestGrammarCheckingOfString:(NSString *)stringToCheck range:(NSRange)range language:(NSString *)language options:(NSDictionary<NSString *, id> *)options completionHandler:(void (^)(NSInteger sequenceNumber, NSArray<NSTextCheckingResult *> *results))completionHandler
+{
+    auto overrideResult = retainPtr([_results objectForKey:stringToCheck]);
```
```suggestion
    RetainPtr overrideResult = retainPtr([_results objectForKey:stringToCheck]);
```

aprotyas reviewed (approved):
Looks good to me with a few questions/suggestions. I think you'll have to compile out a few more things to get the GTK build to pass.

pxlcoder commented on Source/WebCore/editing/Editor.cpp:
```diff
@@ -3208,6 +3208,20 @@ void Editor::markAllMisspellingsAndBadGrammarInRanges(OptionSet<TextCheckingType
+    static bool extendedProofreadingEnabled = false;
+    static std::once_flag flag;
+    std::call_once(flag, [this] {
+        extendedProofreadingEnabled = document().settings().textRecognitionInVideosEnabled();
```
This doesn't seem right ("textRecognitionInVideosEnabled").

pxlcoder replied:
I also don't think we need to cache the value at all. The setting could be toggled.

aprotyas replied:
btw we also have thread-safe statics enabled in WebCore now, so if we do proceed with the caching you can drop the `std::call_once` machinery

```
static bool extendedProofreadingEnabled = document().settings().textRecognitionInVideosEnabled();
```

pxlcoder commented on Source/WebCore/editing/Editor.cpp:
```diff
@@ -3208,6 +3208,20 @@ void Editor::markAllMisspellingsAndBadGrammarInRanges(OptionSet<TextCheckingType
     checkTextOfParagraph(*textChecker(), paragraphToCheck.text(), resolvedOptions, results, document().selection().selection());
     markAndReplaceFor(request.releaseNonNull(), results);
+
+#if PLATFORM(IOS_FAMILY) || PLATFORM(MAC)
```
```suggestion
#if PLATFORM(COCOA)
```

pxlcoder commented on Source/WebKit/Shared/Cocoa/WebPreferencesDefaultValuesCocoa.mm:
```diff
@@ -78,6 +78,16 @@ bool defaultScrollAnimatorEnabled()
 
+bool defaultExtendedProofreadingEnabled()
+{
+#if PLATFORM(MAC) || PLATFORM(IOS_FAMILY)
```
This file is `PLATFORM(COCOA)`, so the guard is unnecessary.

pxlcoder commented on Source/WebCore/editing/SpellChecker.h:
```diff
@@ -110,6 +113,7 @@ class SpellChecker : public CanMakeSingleThreadWeakPtr<SpellChecker> {
 
     RefPtr<SpellCheckRequest> m_processingRequest;
     Deque<Ref<SpellCheckRequest>> m_requestQueue;
+    bool m_inRecheck;
```
```suggestion
    bool m_inRecheck { false };
```

pxlcoder commented on LayoutTests/editing/input/cocoa/extended-proofreading.html:
```diff
@@ -0,0 +1,65 @@
+            testRunner.notifyDone();
+        });
+    }
+    
```
```suggestion
```

pxlcoder commented on LayoutTests/editing/input/cocoa/extended-proofreading.html:
```diff
@@ -0,0 +1,65 @@
+}
+</script>
+<body>
+<script src="../../../resources/js-test-pre.js"></script>
```
Use `js-test.js` and move the include up next to ui-helper.

pxlcoder commented on LayoutTests/editing/input/cocoa/extended-proofreading.html:
```diff
@@ -0,0 +1,65 @@
+
+<script>
+if (window.testRunner)
+    testRunner.waitUntilDone();
```
`jsTestIsAsync = true;`, once using js-test.js

pxlcoder commented on LayoutTests/editing/input/cocoa/extended-proofreading.html:
```diff
@@ -0,0 +1,65 @@
+
+    if (window.testRunner) {
+        shouldBecomeEqual('internals.hasGrammarMarker(15, 2)', 'true', function() {
+            testRunner.notifyDone();
```
`finishJSTest();`, once using js-test.js

pxlcoder commented on LayoutTests/editing/input/cocoa/extended-proofreading.html:
```diff
@@ -0,0 +1,65 @@
+<script>
+if (window.testRunner) {
+    internals.settings.setExtendedProofreadingEnabled(true);
+    testRunner.dumpAsText();
```
`testRunner.dumpAsText` won't be necessary once using `js-test.js` correctly.

hortont424 commented on LayoutTests/editing/input/cocoa/extended-proofreading.html:
```diff
@@ -0,0 +1,59 @@
+
+if (window.internals) {
+    internals.setContinuousSpellCheckingEnabled(true);
+    internals.settings.setExtendedProofreadingEnabled(true);
```
Put this back in the header but capitalize it as ExtendedProofreadingEnabled and it will work. (I'll go look and see if I can fix that)

megangardner committed:
Mac support for extended proofreading.
https://bugs.webkit.org/show_bug.cgi?id=302917
rdar://165075699

Reviewed by Aditya Keerthi and Abrar Rahman Protyasha.

Add support for extended proofreading.

Test: editing/input/cocoa/extended-proofreading.html
* LayoutTests/editing/input/cocoa/extended-proofreading-expected.txt: Added.
* LayoutTests/editing/input/cocoa/extended-proofreading.html: Added.
* LayoutTests/platform/ios/TestExpectations:
* Source/WebCore/PAL/pal/spi/mac/NSSpellCheckerSPI.h:
* Source/WebCore/editing/Editor.cpp:
(WebCore::Editor::markAllMisspellingsAndBadGrammarInRanges):
* Source/WebCore/editing/SpellChecker.cpp:
(WebCore::SpellCheckRequest::didSucceed):
(WebCore::SpellCheckRequest::setExistingResults):
(WebCore::SpellChecker::requestExtendedCheckingFor):
(WebCore::containsGrammarResult):
(WebCore::containsAdditionalGrammarResults):
(WebCore::SpellChecker::didCheck):
(WebCore::SpellChecker::didCheckSucceed):
(WebCore::SpellChecker::didCheckCancel):
* Source/WebCore/editing/SpellChecker.h:
* Source/WebCore/loader/EmptyClients.cpp:
(WebCore::EmptyEditorClient::EmptyTextCheckerClient::requestExtendedCheckingOfString):
* Source/WebCore/platform/text/TextCheckerClient.h:
* Source/WebKit/UIProcess/TextChecker.cpp:
(WebKit::TextChecker::requestExtendedCheckingOfString):
* Source/WebKit/UIProcess/TextChecker.h:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::requestExtendedCheckingOfString):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebPageProxy.messages.in:
* Source/WebKit/UIProcess/ios/TextCheckerIOS.mm:
(WebKit::TextChecker::requestExtendedCheckingOfString):
* Source/WebKit/UIProcess/mac/TextCheckerMac.mm:
(WebKit::convertExtendedCheckingResults):
(WebKit::TextChecker::requestExtendedCheckingOfString):
* Source/WebKit/WebProcess/WebCoreSupport/WebEditorClient.cpp:
(WebKit::WebEditorClient::requestExtendedCheckingOfString):
* Source/WebKit/WebProcess/WebCoreSupport/WebEditorClient.h:
* Source/WebKitLegacy/mac/WebCoreSupport/WebEditorClient.h:
* Source/WebKitLegacy/mac/WebCoreSupport/WebEditorClient.mm:
(WebEditorClient::requestExtendedCheckingOfString):
* Tools/TestRunnerShared/cocoa/LayoutTestSpellChecker.mm:
(-[LayoutTestSpellChecker requestGrammarCheckingOfString:range:language:options:completionHandler:]):

Canonical link: https://commits.webkit.org/303913@main
